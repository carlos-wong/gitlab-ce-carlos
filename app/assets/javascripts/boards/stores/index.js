import Vue from 'vue';
import Vuex from 'vuex';
import state from 'ee_else_ce/boards/stores/state';
import getters from 'ee_else_ce/boards/stores/getters';
import actions from 'ee_else_ce/boards/stores/actions';
import mutations from 'ee_else_ce/boards/stores/mutations';

Vue.use(Vuex);

export const createStore = () =>
  new Vuex.Store({
    state,
    getters,
    actions,
    mutations,
  });

export default createStore();
