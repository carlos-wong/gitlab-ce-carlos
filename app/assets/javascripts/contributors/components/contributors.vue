<script>
import _ from 'underscore';
import { mapActions, mapState, mapGetters } from 'vuex';
import { GlLoadingIcon } from '@gitlab/ui';
import { GlAreaChart } from '@gitlab/ui/dist/charts';
import { __ } from '~/locale';
import { getSvgIconPathContent } from '~/lib/utils/icon_utils';
import { getDatesInRange } from '~/lib/utils/datetime_utility';
import { xAxisLabelFormatter, dateFormatter } from '../utils';

export default {
  components: {
    GlAreaChart,
    GlLoadingIcon,
  },
  props: {
    endpoint: {
      type: String,
      required: true,
    },
    branch: {
      type: String,
      required: true,
    },
  },
  data() {
    return {
      masterChart: null,
      individualCharts: [],
      svgs: {},
      masterChartHeight: 264,
      individualChartHeight: 216,
    };
  },
  computed: {
    ...mapState(['chartData', 'loading']),
    ...mapGetters(['showChart', 'parsedData']),
    masterChartData() {
      const data = {};
      this.xAxisRange.forEach(date => {
        data[date] = this.parsedData.total[date] || 0;
      });
      return [
        {
          name: __('Commits'),
          data: Object.entries(data),
        },
      ];
    },
    masterChartOptions() {
      return {
        ...this.getCommonChartOptions(true),
        yAxis: {
          name: __('Number of commits'),
        },
        grid: {
          bottom: 64,
          left: 64,
          right: 20,
          top: 20,
        },
      };
    },
    individualChartsData() {
      const maxNumberOfIndividualContributorsCharts = 100;

      return Object.keys(this.parsedData.byAuthor)
        .map(name => {
          const author = this.parsedData.byAuthor[name];
          return {
            name,
            email: author.email,
            commits: author.commits,
            dates: [
              {
                name: __('Commits'),
                data: this.xAxisRange.map(date => [date, author.dates[date] || 0]),
              },
            ],
          };
        })
        .sort((a, b) => b.commits - a.commits)
        .slice(0, maxNumberOfIndividualContributorsCharts);
    },
    individualChartOptions() {
      return {
        ...this.getCommonChartOptions(false),
        yAxis: {
          name: __('Commits'),
          max: this.individualChartYAxisMax,
        },
        grid: {
          bottom: 27,
          left: 64,
          right: 20,
          top: 8,
        },
      };
    },
    individualChartYAxisMax() {
      return this.individualChartsData.reduce((acc, item) => {
        const values = item.dates[0].data.map(value => value[1]);
        return Math.max(acc, ...values);
      }, 0);
    },
    xAxisRange() {
      const dates = Object.keys(this.parsedData.total).sort((a, b) => new Date(a) - new Date(b));

      const firstContributionDate = new Date(dates[0]);
      const lastContributionDate = new Date(dates[dates.length - 1]);

      return getDatesInRange(firstContributionDate, lastContributionDate, dateFormatter);
    },
    firstContributionDate() {
      return this.xAxisRange[0];
    },
    lastContributionDate() {
      return this.xAxisRange[this.xAxisRange.length - 1];
    },
    charts() {
      return _.uniq(this.individualCharts);
    },
  },
  mounted() {
    this.fetchChartData(this.endpoint);
  },
  methods: {
    ...mapActions(['fetchChartData']),
    getCommonChartOptions(isMasterChart) {
      return {
        xAxis: {
          type: 'time',
          name: '',
          data: this.xAxisRange,
          axisLabel: {
            formatter: xAxisLabelFormatter,
            showMaxLabel: false,
            showMinLabel: false,
          },
          boundaryGap: false,
          splitNumber: isMasterChart ? 24 : 18,
          // 28 days
          minInterval: 28 * 86400 * 1000,
          min: this.firstContributionDate,
          max: this.lastContributionDate,
        },
      };
    },
    setSvg(name) {
      return getSvgIconPathContent(name)
        .then(path => {
          if (path) {
            this.$set(this.svgs, name, `path://${path}`);
          }
        })
        .catch(() => {});
    },
    onMasterChartCreated(chart) {
      this.masterChart = chart;
      this.setSvg('scroll-handle')
        .then(() => {
          this.masterChart.setOption({
            dataZoom: [
              {
                type: 'slider',
                handleIcon: this.svgs['scroll-handle'],
              },
            ],
          });
        })
        .catch(() => {});
      this.masterChart.on('datazoom', _.debounce(this.setIndividualChartsZoom, 200));
    },
    onIndividualChartCreated(chart) {
      this.individualCharts.push(chart);
    },
    setIndividualChartsZoom(options) {
      this.charts.forEach(chart =>
        chart.setOption(
          {
            dataZoom: {
              start: options.start,
              end: options.end,
              show: false,
            },
          },
          { lazyUpdate: true },
        ),
      );
    },
  },
};
</script>

<template>
  <div>
    <div v-if="loading" class="contributors-loader text-center">
      <gl-loading-icon :inline="true" :size="4" />
    </div>

    <div v-else-if="showChart" class="contributors-charts">
      <h4>{{ __('Commits to') }} {{ branch }}</h4>
      <span>{{ __('Excluding merge commits. Limited to 6,000 commits.') }}</span>
      <div>
        <gl-area-chart
          :data="masterChartData"
          :option="masterChartOptions"
          :height="masterChartHeight"
          @created="onMasterChartCreated"
        />
      </div>

      <div class="row">
        <div v-for="contributor in individualChartsData" :key="contributor.name" class="col-6">
          <h4>{{ contributor.name }}</h4>
          <p>{{ n__('%d commit', '%d commits', contributor.commits) }} ({{ contributor.email }})</p>
          <gl-area-chart
            :data="contributor.dates"
            :option="individualChartOptions"
            :height="individualChartHeight"
            @created="onIndividualChartCreated"
          />
        </div>
      </div>
    </div>
  </div>
</template>
