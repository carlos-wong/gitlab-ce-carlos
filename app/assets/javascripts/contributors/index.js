import Vue from 'vue';
import ContributorsGraphs from './components/contributors.vue';
import store from './stores';

export default () => {
  const el = document.querySelector('.js-contributors-graph');

  if (!el) return null;

  return new Vue({
    el,
    store,

    render(createElement) {
      return createElement(ContributorsGraphs, {
        props: {
          endpoint: el.dataset.projectGraphPath,
          branch: el.dataset.projectBranch,
        },
      });
    },
  });
};
