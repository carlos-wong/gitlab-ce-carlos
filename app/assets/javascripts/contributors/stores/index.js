import Vue from 'vue';
import Vuex from 'vuex';
import state from './state';
import mutations from './mutations';
import * as getters from './getters';
import * as actions from './actions';

Vue.use(Vuex);

export const createStore = () =>
  new Vuex.Store({
    actions,
    mutations,
    getters,
    state: state(),
  });

export default createStore();
