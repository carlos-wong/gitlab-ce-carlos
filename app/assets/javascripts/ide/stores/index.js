import Vue from 'vue';
import Vuex from 'vuex';
import state from './state';
import * as actions from './actions';
import * as getters from './getters';
import mutations from './mutations';
import commitModule from './modules/commit';
import pipelines from './modules/pipelines';
import mergeRequests from './modules/merge_requests';
import branches from './modules/branches';
import fileTemplates from './modules/file_templates';
import paneModule from './modules/pane';
import clientsideModule from './modules/clientside';

Vue.use(Vuex);

export const createStore = () =>
  new Vuex.Store({
    state: state(),
    actions,
    mutations,
    getters,
    modules: {
      commit: commitModule,
      pipelines,
      mergeRequests,
      branches,
      fileTemplates: fileTemplates(),
      rightPane: paneModule(),
      clientside: clientsideModule(),
    },
  });

export default createStore();
