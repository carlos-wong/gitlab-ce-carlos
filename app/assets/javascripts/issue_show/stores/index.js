import _ from 'underscore';
import { convertObjectPropsToCamelCase } from '~/lib/utils/common_utils';
import updateDescription from '../utils/update_description';

export default class Store {
  constructor(initialState) {
    this.state = initialState;
    this.formState = {
      title: '',
      description: '',
      lockedWarningVisible: false,
      updateLoading: false,
      lock_version: 0,
      issuableTemplates: [],
    };
  }

  updateState(data) {
    if (this.stateShouldUpdate(data)) {
      this.formState.lockedWarningVisible = true;
    }

    Object.assign(this.state, convertObjectPropsToCamelCase(data));
    // find if there is an open details node inside of the issue description.
    const descriptionSection = document.body.querySelector(
      '.detail-page-description.content-block',
    );
    const details =
      !_.isNull(descriptionSection) && descriptionSection.getElementsByTagName('details');

    this.state.descriptionHtml = updateDescription(data.description, details);
    this.state.titleHtml = data.title;
    this.state.lock_version = data.lock_version;
  }

  stateShouldUpdate(data) {
    return (
      this.state.titleText !== data.title_text ||
      this.state.descriptionText !== data.description_text
    );
  }

  setFormState(state) {
    this.formState = Object.assign(this.formState, state);
  }
}
