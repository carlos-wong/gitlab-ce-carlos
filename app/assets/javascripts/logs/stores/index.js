import Vue from 'vue';
import Vuex from 'vuex';
import * as actions from './actions';
import * as getters from './getters';
import mutations from './mutations';
import state from './state';

Vue.use(Vuex);

export const createStore = () =>
  new Vuex.Store({
    modules: {
      environmentLogs: {
        namespaced: true,
        actions,
        mutations,
        state: state(),
        getters,
      },
    },
  });

export default createStore;
