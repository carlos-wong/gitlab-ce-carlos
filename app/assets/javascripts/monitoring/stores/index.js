import Vue from 'vue';
import Vuex from 'vuex';
import * as actions from './actions';
import * as getters from './getters';
import mutations from './mutations';
import state from './state';

Vue.use(Vuex);

export const createStore = () =>
  new Vuex.Store({
    modules: {
      monitoringDashboard: {
        namespaced: true,
        actions,
        getters,
        mutations,
        state,
      },
    },
  });

export default createStore();
