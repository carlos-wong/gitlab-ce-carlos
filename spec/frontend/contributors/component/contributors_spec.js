import Vue from 'vue';
import { mount } from '@vue/test-utils';
import MockAdapter from 'axios-mock-adapter';
import { createStore } from '~/contributors/stores';
import axios from '~/lib/utils/axios_utils';
import ContributorsCharts from '~/contributors/components/contributors.vue';

let wrapper;
let mock;
let store;
const Component = Vue.extend(ContributorsCharts);
const endpoint = 'contributors';
const branch = 'master';
const chartData = [
  { author_name: 'John', author_email: 'jawnnypoo@gmail.com', date: '2019-05-05' },
  { author_name: 'John', author_email: 'jawnnypoo@gmail.com', date: '2019-03-03' },
];

function factory() {
  mock = new MockAdapter(axios);
  jest.spyOn(axios, 'get');
  mock.onGet().reply(200, chartData);
  store = createStore();

  wrapper = mount(Component, {
    propsData: {
      endpoint,
      branch,
    },
    stubs: {
      GlLoadingIcon: true,
      GlAreaChart: true,
    },
    store,
  });
}

describe('Contributors charts', () => {
  beforeEach(() => {
    factory();
  });

  afterEach(() => {
    mock.restore();
    wrapper.destroy();
  });

  it('should fetch chart data when mounted', () => {
    expect(axios.get).toHaveBeenCalledWith(endpoint);
  });

  it('should display loader whiled loading data', () => {
    wrapper.vm.$store.state.loading = true;
    return wrapper.vm.$nextTick(() => {
      expect(wrapper.find('.contributors-loader').exists()).toBe(true);
    });
  });

  it('should render charts when loading completed and there is chart data', () => {
    wrapper.vm.$store.state.loading = false;
    wrapper.vm.$store.state.chartData = chartData;
    return wrapper.vm.$nextTick(() => {
      expect(wrapper.find('.contributors-loader').exists()).toBe(false);
      expect(wrapper.find('.contributors-charts').exists()).toBe(true);
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
