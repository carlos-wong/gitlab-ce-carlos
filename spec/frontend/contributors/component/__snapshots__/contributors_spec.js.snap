// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contributors charts should render charts when loading completed and there is chart data 1`] = `
<div>
  <div
    class="contributors-charts"
  >
    <h4>
      Commits to master
    </h4>
     
    <span>
      Excluding merge commits. Limited to 6,000 commits.
    </span>
     
    <div>
      <glareachart-stub
        data="[object Object]"
        height="264"
        includelegendavgmax="true"
        legendaveragetext="Avg"
        legendmaxtext="Max"
        option="[object Object]"
        thresholds=""
        width="0"
      />
    </div>
     
    <div
      class="row"
    >
      <div
        class="col-lg-6 col-12"
      >
        <h4>
          John
        </h4>
         
        <p>
          2 commits (jawnnypoo@gmail.com)
        </p>
         
        <div>
          <glareachart-stub
            data="[object Object]"
            height="216"
            includelegendavgmax="true"
            legendaveragetext="Avg"
            legendmaxtext="Max"
            option="[object Object]"
            thresholds=""
            width="0"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
